### DISCUSSION TITLE
cannot write data in loop_function

### DISCUSSION THREAD

**Discussion by Qiuishere** [jspsych/jsPsych#2580] [Q&A] [answered]
Hi!

I am using loop_function to achieve that everytime you press uparrow or downarrow, the image will change, and when you press space, the loop will be ended. The presentation works well, but I got trouble with the data writing. Basically, I'm not able to write data into the datafile. Here is the code:

```
 const adjustment = {
        type: 'image-keyboard-response',
        stimulus: function(data){
            if (jsPsych.data.get().last(1).values()[0].stimulus == 'stimuli/masks/' + maskimg[Nmask-1] + '.jpg'){
                initImg = jsPsych.data.get().last(5).values()[0].stimulus;
                initAng = parseInt(initImg.match(/\d+/g));
                theang = initAng;
                currentImg = initImg;
            } else {
                currentImg = initImg.replace(initAng,theang)
            }
            return currentImg;
        },
        stimulus_width: imgSize,
        response_type: 'key',
        choices: [' ', 'ArrowUp', 'ArrowDown'],
    };

    const loop_node = {
        timeline: [adjustment],
        data: jsPsych.timelineVariable('data'),

        loop_function: function(data){
            if (data.values()[0].response == 'arrowup'){
                theang -= 3;
                return true

            } else if (data.values()[0].response == 'arrowdown'){ // arrowdown
                theang += 3;
                return true

            } else if (data.values()[0].response == ' '){
                data.respAng = theang;
                return false
            }
        },

    }
aTrial.push(loop_node)
```

At the end of this loop, i.e. when the subject presses space, I would like to store the current value "theang" into the data file. However, nothing apprears in the data file. I also tried to put 'data.respAng = theang;' in the above branches of if structure, where the condition leads to return true, also didn't work. 

If I put data in the on_finish parameter as for other normal trials, the trial will be ended when you press a key, i.e. no loop at all. 
`        on_finish: function(data) {
            data.respAng = theang
        },`

Then I tried the on_timeline_finish, as this:
`        on_timeline_finish: function(data){
            data.respAng = theang;
        }`
Then it reported an error:
_index.html:396 Uncaught TypeError: Cannot set properties of undefined (setting 'respAng')_

I guess the loop_function does not allow writing data except those already stored in the **data** object created inside the loop? 

My question is how to write the properties that I'd like to store in my data file?
An extra question is why putting an on_finish will render the loop trial an normal non-loop trial?

Thanks so much if anyone has any idea how to achieve this, or provide more information about how the data object works inside the loop_function.

Good day!
Qiu

**Comment by jodeleeuw** [comment, marked as answer]
Hi @Qiuishere,

> I guess the loop_function does not allow writing data except those already stored in the data object created inside the loop?

Yes, that's correct. This is because the data returned from the loop function is the `DataCollection` for all trials executed inside the loop. You could, I think, access individual trials within that collection and modify them.

> My question is how to write the properties that I'd like to store in my data file?

I would recommend adding the `on_finish` to the `adjustment` trial.

```js
on_finish: functon(data){
	if(data.response == " "){
		data.respAng = theang;
	}
}
```

> An extra question is why putting an on_finish will render the loop trial an normal non-loop trial?

I'm guessing that what happened is that the `on_finish` on the `loop_node` created an error because you can't assign `data.respAng` so the script halted. But, please let us know if the loop doesn't work when adding the `on_finish` to the `adjustment` trial. I think that would indicate a bug that needs to be fixed.

**Reply by Qiuishere** [reply]
Hi Josh,

Thanks very much! 

> I would recommend adding the on_finish to the adjustment trial.

This indeed worked well. Thanks again!

> But, please let us know if the loop doesn't work when adding the on_finish to the adjustment trial.

To make sure I tried to put `on_finish: function(data) { data.respAng = theang },` inside `adjustment` , actually it worked well, even without the `if(data.response == " ")` condition. So I guess it's not a problem.

This problem is well solved! May I ask another question for this same script about image presentation? 
When I press uparrow/downarrow to change the picture, the flickering of refreshing is quite obvious. I'm running locally with Jspsych 6.3. I think this might have something to do with preload: e.g. if I press uparrow to change img1 to img2, I will see a clear flickering, then I press downarrow to go back to img1, there is no flickering. Then if I change to a new image, flickering occurs again. As if img2 has already been loaded and stored in memory. But this also lasts for a very short time: after a few seconds if I change to img2 again, the flickering might appear again (but not always).
The problem is I indeed preloaded all the images used, either by getting all the images in all the trials to preload:

```
    var imgstopreload = trialArray.map(element => element.IMG)
    var preload = {
        type: "preload",
        images: imgstopreload,
        auto_preload: true,
        show_detailed_errors: true
    };
    timeline.push(preload);
```
where `trialArray `contains all the trial used. `trialArray.IMG` contains the first image of the loop. But all the images used in the adjustment phase will be covered since they could be the first image of other trials.

Or use the automatic preload:

```
    var preload = {
        type: 'preload',
        auto_preload: true
    }
    timeline.push(preload);
```

Neither of these could solve the problem.
One reason I can think of is maybe we have to preload all the images used in one trial simultaneously? i.e. for each trial we need to include all the possible images that the subject might adjust to (which means almost all the images), and push them together?

Thanks for any help! Have a good weekend!
Qiu
